File: js/src/http/response.ts
import { STATUS_CODES } from 'http'
import { Buffer } from 'buffer'

import { encode } from '../encode.js'

const version = 1.1,
    contentType = {
        text: 'text/plain',
        other: 'application/octet-stream'
    }

export class Response {
    constructor(private code: number | string, private payload?) {
        if (typeof payload === 'string') this._contentType = contentType.text
        else this._contentType = contentType.other
    }

    private _contentType: string
    get contentType() { return this._contentType }
    
    private _encoding: string
    private payloadEncoded
    set encoding(encoding: string) { 
        this._encoding = encoding 
        this.payloadEncoded = encode(this.payload, encoding)
    }

    /** The Content-Length of the response message, in bytes */
    length() { return Buffer.byteLength(this._encoding ? this.payloadEncoded : this.payload)}

    headers() {
        const headers = [`HTTP/${version} ${this.code} ${STATUS_CODES[this.code]}`,
            'Content-Type: ' + this._contentType,
            'Content-Length: ' + this.length()]
        if (this._encoding) headers.push('Content-Encoding: ' + this._encoding)

        headers.push('', '') // Header-body separator
        return headers.join('\r\n')
    }
    payloadBuffer() {
        let payload = this.encoding ? this.payloadEncoded : this.payload

        if (this._encoding) payload = encode(payload, this._encoding)
        let buffer = Buffer.isBuffer(payload) ? payload : Buffer.from(payload)
        // if (this._encoding) buffer = encode(buffer, this._encoding)              // Encoding before or after buffering results in the same

        return buffer
    }
}

File: src/http/response.cpp
#include "response.hpp"

#include <sys/socket.h>
#include <vector>

#include "../utils/console.hpp"
#include "../utils/utils.hpp"

using std::string;
using std::to_string;

const string VERSION = "1.1";

//                             Pointer to this occurence's code in memory
Response::Response(int code) { this->code = code; }
Response::Response(int code, string message) {
    this->code = code;
    content_type = "text/plain";
    this->message = message;
    file = nullptr;
}
Response::Response(int code, std::ifstream *file) {
    this->code = code;
    content_type = "application/octet-stream";
    this->file = file;
}

string Response::headers() {
    std::vector<string> headers_vector;
    headers_vector.push_back("HTTP/" + VERSION + " " + to_string(code) + " " + status());
    if (!content_type.empty()) headers_vector.push_back("Content-Type: " + content_type);
    if (!message.empty()) headers_vector.push_back("Content-Length: " + to_string(this->message.size()));
    if (file != nullptr && file->is_open()) {
        file->seekg(0, std::ios::end); // Move the file pointer to the end
        std::streampos fileSize = file->tellg();
        headers_vector.push_back("Content-Length: " + to_string(fileSize));
        file->seekg(0, std::ios::beg); // Move back to the beginning to read it
    }

    // Header-body separator
    headers_vector.push_back("");
    headers_vector.push_back("");

    return vec::join(headers_vector, "\r\n");
}

int Response::get_code() { return code; }
string Response::msg() { return message; }

// PRIVATE
// Use map instead?
string Response::status() {
    switch (code) {
        case 200:
            return "OK";
        case 400:
            return "Bad Request";
        case 404:
            return "Not Found";
        case 503:
            return "Service Unavailable";
        default:
            return to_string(code);
    }
}
